proxy/vm.go

```diff
@@ -0,0 +1,460 @@
+	// we've just written
+	acked := 0
+	for acked < length {
+		next, err := vm.readMessage()
+		if err != nil {
+			return nil
+		}
+		if next.Code != hyper.INIT_NEXT {
+			return errors.New("didn't receive NEXT from hyperstart")
+		}
+		if len(next.Message) != 4 {
+			return errors.New("wrong size for NEXT message")
+		}
+		acked += int(binary.BigEndian.Uint32(next.Message[0:4]))
+	}
+
+	// Somehow hyperstart acked more than it should?
+	if acked != length {
+		fmt.Fprintf(os.Stderr,
+			"wrong acked size by NEXT message (expected %d, got %d)\n", length, acked)
+	}
+
+	return nil
+}
+
+func (vm *vm) waitForReady() error {
+	vm.ctlMutex.Lock()
+	defer vm.ctlMutex.Unlock()
+
+	msg, err := vm.readMessage()
+	if err != nil {
+		return err
+	} else if msg.Code != hyper.INIT_READY {
+		return fmt.Errorf("Unexpected message %d", msg.Code)
+	}
+
+	return nil
+}
+
+// Returns one chunk of data belonging to the seq stream
+func readIoMessage(conn net.Conn) (seq uint64, data []byte, err error) {
+	// see runv's hypervisor/tty.go
+	needRead := 12
+	length := 0
+	read := 0
+	buf := make([]byte, 512)
+	res := []byte{}
+	for read < needRead {
+		want := needRead - read
+		if want > 512 {
+			want = 512
+		}
+		nr, err := conn.Read(buf[:want])
+		if err != nil {
+			return 0, nil, err
+		}
+
+		res = append(res, buf[:nr]...)
+		read = read + nr
+
+		if length == 0 && read >= 12 {
+			length = int(binary.BigEndian.Uint32(res[8:12]))
+			if length > 12 {
+				needRead = length
+			}
+		}
+	}
+
+	seq = binary.BigEndian.Uint64(res[:8])
+	return seq, res, nil
+
+}
+
+func (vm *vm) findClientConn(seq uint64) net.Conn {
+	vm.Lock()
+	defer vm.Unlock()
+
+	session := vm.ioSessions[seq]
+	if session == nil {
+		return nil
+	}
+	return vm.ioSessions[seq].client
+}
+
+// This function runs in a goroutine, reading data from the io channel and
+// dispatching it to the right client (the one with matching seq number)
+// There's only one instance of this goroutine per-VM
+func (vm *vm) ioHyperToClients() {
+	for {
+		seq, data, err := readIoMessage(vm.io)
+		if err != nil {
+			// VM process is gone
+			break
+		}
+
+		client := vm.findClientConn(seq)
+		if client == nil {
+			fmt.Fprintf(os.Stderr,
+				"couldn't find client with seq number %d\n", seq)
+			continue
+		}
+		n, err := client.Write(data)
+		if err != nil || n != len(data) {
+			fmt.Fprintf(os.Stderr,
+				"error writing I/O data to client: %v (%d bytes written)\n", err, n)
+
+			break
+
+		}
+	}
+
+	vm.wg.Done()
+}
+
+func (vm *vm) Connect() error {
+	if err := vm.openSockets(); err != nil {
+		return err
+	}
+	if err := vm.waitForReady(); err != nil {
+		vm.closeSockets()
+		vm.ctl = nil
+		vm.io = nil
+		return err
+	}
+
+	vm.wg.Add(1)
+	go vm.ioHyperToClients()
+
+	return nil
+}
+
+func cmdFromSring(cmd string) (uint32, error) {
+	// Commands not supported by proxy:
+	//   hyper.INIT_STOPPOD_DEPRECATED
+	//   hyper.INIT_WRITEFILE
+	//   hyper.INIT_READFILE
+	switch cmd {
+	case "version":
```

Comment:
Maybe, but there are commands we can't have in there as we can't wrap them in JSON (writefile/readfile), so the autogeneration would have to know that. Seems like a lot of work for very little gain.